core/src/main/java/fr/sncf/osrd/train/TrainPositionTracker.java

```diff
@@ -104,251 +128,41 @@ public TrainPositionTracker clone() {
-    }
-
-    private TrackSectionRange previousTrackSectionPosition(double delta) {
-        var curTrackSectionPos = trackSectionRanges.getLast();
-        var neighbors = infra.trackGraph.getStartNeighborRels(
-                curTrackSectionPos.edge,
-                curTrackSectionPos.direction
-        );
-
-        if (neighbors.isEmpty())
-            return null;
-
-        var prev = neighbors.get(neighbors.size() - 1);
-        var prevTrackSection = prev.getEdge(curTrackSectionPos.edge, curTrackSectionPos.direction);
-
-        // In case of a switch, we need to get the next track section align with the position of the switch.
-        if (neighbors.size() > 1) {
-            assert ignoreInfraState;
-            // If we need to ignore the infra state, we refer to the given path instead
-            prevTrackSection = null;
-            var currentEdge = curTrackSectionPos.edge;
-            for (int i = 1; i < trackSectionPath.size(); i++) {
-                if (trackSectionPath.get(i).edge.id.equals(currentEdge.id)) {
-                    prevTrackSection = trackSectionPath.get(i - 1).edge;
-                    break;
-                }
-            }
-            if (prevTrackSection == null)
-                throw new RuntimeException("Can't move train further because it has reached the end of its path");
-        }
-
-        var prevTrackSectionDirection = prevTrackSection.getDirection(
-                curTrackSectionPos.edge, curTrackSectionPos.direction.opposite()).opposite();
-        return TrackSectionRange.makePrev(prevTrackSection, prevTrackSectionDirection, delta);
+    public TrackSectionLocation getHeadLocation() {
+        return TrainPath.findLocation(headPathPosition, trackSectionPath);
     }
 
     /**
      * Updates the position of the train on the network.
      * @param positionDelta How much the train moves by.
      */
     public void updatePosition(double expectedTrainLength, double positionDelta) {
-        if (positionDelta >= 0)
-            updateHeadPosition(positionDelta);
-        else
-            updateHeadPositionBackwards(positionDelta);
-        pathPosition += positionDelta;
-
-        double currentTrainLength = trackSectionRanges.stream().mapToDouble(Range::length).sum();
-
-        expectedTrainLength = min(expectedTrainLength, pathPosition);
-
-        var tailDisplacement = currentTrainLength - expectedTrainLength;
-        if (tailDisplacement > 0 && positionDelta > 0)
-            updateTailPosition(tailDisplacement);
-        else if (tailDisplacement < 0 && positionDelta < 0)
-            updateTailPositionBackwards(tailDisplacement);
-
-        assertIntegrity();
-        var actualTrainLength = trackSectionRanges.stream().mapToDouble(Range::length).sum();
-        assert abs(actualTrainLength - expectedTrainLength) < 1e-3;
+        headPathPosition += positionDelta;
+        tailPathPosition = headPathPosition - expectedTrainLength;
```

Comment:
It seems impressive that these two lines work instead of all the other methods for updating head and tail, but i guess the work you did around make it possible ?
Does it properly take into account the negative positionDeltas ?